src/AsyncBridge/AsyncCompatLibExtensions.cs

```diff
@@ -168,29 +168,28 @@ public static void CancelAfter(this CancellationTokenSource cancelSource, TimeSp
         if (cancelSource.IsCancellationRequested)
             return;
 
-        Timer MyTimer = null;
+        Timer myTimer = null;
 
         // CTS claims it's thread-safe for all methods, so we should be the same.
         // If someone calls CancelAfter concurrently for the same CTS, we won't explode or cause a memory leak
-        do
+        while (!_cancelTimers.TryGetTimer(cancelSource, out myTimer))
         {
-            if (_cancelTimers.TryGetTimer(cancelSource, out var CurrentTimer))
-            {
-                if (MyTimer != null)
-                    MyTimer.Dispose();
-
-                MyTimer = CurrentTimer;
-                break; // Reset the existing timer to the new delay
-            }
-
             // An active timer doesn't hold a strong reference to itself, so adding the CTS as state won't hold it alive
-            MyTimer = new Timer(OnCancelAfterTimer, cancelSource, Timeout.Infinite, Timeout.Infinite);
+            myTimer = new Timer(OnCancelAfterTimer, cancelSource, Timeout.Infinite, Timeout.Infinite);
 
-        } while (!_cancelTimers.TryAddTimer(cancelSource, MyTimer));
+            if (_cancelTimers.TryAddTimer(cancelSource, myTimer))
+                break;
+
+            // TryAddTimer can only fail if a Timer was created concurrently. Since we never remove them, TryGetTimer is guaranteed to succeed
+            // Thus, we can dispose of the new Timer we just created and loop back.
+            // If a thread abort or other exception occurs here, we leave the orphan timer to be GC'ed
+            myTimer.Dispose();
```

Comment:
What's the downside to disposing in a `finally` block, just in case of a thread abort or if `TryAddTimer` ever threw an exception in the future?

Reply:
You really want this to have a `finally` block, eh? :grin:

In regards to `ThreadAbortException`, the rest of the code is not `Thread.Abort` safe, and there's nothing we can do about that. If there's a `ThreadAbortException` (or `OutOfMemoryException`) during one of `ConditionalWeakTable`'s operations, the entire class will mark its state as invalid and will throw an exception whenever you call any of its methods. At that point, you can only restart the AppDomain, since it's a static field. I've no idea what `ConcurrentDictionary` will do, but it's likely just as bad. It's not worth adding code here to protect against it, when our dependencies don't offer the same guarantee.

As for `TryAddTimer` throwing an exception, the only place it can is on 3.5 when we call `CancellationToken.Register` - it can throw `ObjectDisposedException`. The signature allows for this, so it won't be a surprise. And if the CTS has been disposed, chances are it's going to become available for GC pretty soon, which means the orphan Timer will as well.

It's mostly code complexity, and the fact that any solution is an incomplete one at best. Since the code already relies on the garbage collector cleaning up old Timers, I say we let the exceptional cases do the same.

Reply:
You make a good case. :+1: